core/modules/widgets/widget.js

```diff
@@ -263,12 +285,50 @@ Widget.prototype.getStateQualifier = function(name) {
 };
 
 /*
-Compute the current values of the attributes of the widget. Returns a hashmap of the names of the attributes that have changed
+Make a fake widget with specified variables, suitable for variable lookup in filters
+*/
+Widget.prototype.makeFakeWidgetWithVariables = function(variables) {
```

Comment:
I have some thoughts regarding this design choice:

First, if we stop thinking about this "fake widget" object (instantiating the class `Widget`), and instead try to think of it as a "variable store" or similar (implementing an `IVariableStore` interface), we would arrive at a different design.

As I understand it, what we want is an object that responds to `getVariable()` -- and that is not a `Widget`.

The problem with `makeFakeWidgetWithVariables()` is that it's an instance method, so you still need a `Widget` object to call it -- and you run the risk of getting unexpected results if the code using your fake widget ends up in the "`self.getVariable()`" method!

Just yesterday, I arrived at a similar design myself, but one that doesn't involve the `Widget` class at all. Of course, there's a lot of things about this and TiddlyWiki that I just don't understand, so I may be wrong.

My implementation and some thoughts that align well with my comment here: https://talk.tiddlywiki.org/t/javascript-pass-variable-to-e-g-filtertiddlers/5147

Reply:
I'd be inclined to agree - at the very least if this method was just attached to the class directly and not to the prototype it would be accessible to other code. But if we **know for a fact** that real widget is not required for filters to run the name of the variable could be refactored to something else - I'd personally would call it `variableSource` (I think store implies it must store the variables itself, while source makes no assumptions about where those variables come from). Then we could expose this function instead as a method `$tw.wiki.createVariableSource()`.

But I too am not that familiar with TW's internals.

(Also @cdaven I hope you don't mind me butting in into your PRs and comments, sorry!)

Reply:
I implemented this visibly clunky method for the moment because I want to be able to give more consideration to how far we should go with fixing things properly. As you note, a "variableSource" interface would appear to be a reasonable way to go. However, the challenge to being able to properly encapsulate things is that many widgets directly access the `variables` property, making it hard to change their implementation. So, I'm thinking that this might be a case where we defer fixing it until we're ready to make a version that is not fully backwards compatible.

In the meantime, we could change  the name of the method, and provide a static alternative that doesn't require a widget instance.